actions/setup/action.yml

```diff
@@ -43,20 +43,22 @@ runs:
         check-latest: ${{ inputs.check-latest }}
         cache: false  # cache is handled by separate actions/cache step.
 
+    # NOTE: cache key and restore key is not the same;
+    # We want to always re-use the cache created by master in all our PRs. We also want any PR which modifies go.sum to create a new cache for itself.
+    # Therefore the cache key is set to use GITHUB_REF_NAME and the restore key is set to use GITHUB_BASE_REF.
+    # For more details, see https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables
+    # NOTE: you need to build on push to master and on pull request in the repos which uses this reusable workflow.
     - name: Set up cache
```

Comment:
It's possible that the absolutely best setup for us would be to always only have `master` _write_ the cache. Then we just `read` the cache in all PRs.

However, this would require us to break this workflow up into two parts:

```yml
      - name: Restore cache
        if: ${{ inputs.disableCache != 'true' }}
        uses: actions/cache/restore@v4
        with:
          path: |
            ./.sage/build
            ./.sage/tools
            ./.sage/bin
            /home/runner/.cache/go-build
            /home/runner/go/pkg/mod
            /home/runner/go/bin
          key: ${{ runner.os }}-${{ github.ref_name }}-${{ github.workflow }}-${{ github.job }}-${{ inputs.cacheKey }}-${{ inputs.go-version }}-${{ hashFiles('**/go.sum') }}
          restore-keys: |
            ${{ runner.os }}-${{ github.base_ref }}-${{ github.workflow }}-${{ github.job }}-${{ inputs.cacheKey }}-${{ inputs.go-version }}-${{ hashFiles('**/go.sum') }}

      # Other steps in your job would go here

      - name: Save cache
        if: ${{ inputs.disableCache != 'true' && (github.ref_name == 'master' || github.ref_name == 'main') }}
        uses: actions/cache/save@v4
        with:
          path: |
            ./.sage/build
            ./.sage/tools
            ./.sage/bin
            /home/runner/.cache/go-build
            /home/runner/go/pkg/mod
            /home/runner/go/bin
          key: ${{ runner.os }}-${{ github.ref_name }}-${{ github.workflow }}-${{ github.job }}-${{ inputs.cacheKey }}-${{ inputs.go-version }}-${{ hashFiles('**/go.sum') }}
```

And that would be a much bigger undertaking. So I propose we don't do that as step 1.